chainer/training/extensions/extractor.py

```diff
@@ -0,0 +1,199 @@
+
+    Attributes:
+        converter: Converter function.
+        device: Device to which the training data is sent.
+        extract_hook: Function to prepare for each extraction process.
+        extract_func: Extraction function called at each iteration.
+        features: The extracted features. Created after calling the extension.
+    """
+    trigger = 1, 'epoch'
+    default_name = 'extraction'
+    priority = extension.PRIORITY_WRITER
+
+    def __init__(self, iterator, target, converter=convert.concat_examples,
+                 device=None, extract_hook=None, extract_func=None,
+                 merger=concat_variables):
+        if isinstance(iterator, iterator_module.Iterator):
+            iterator = {'main': iterator}
+        self._iterators = iterator
+
+        if isinstance(target, link.Link):
+            target = {'main': target}
+        self._targets = target
+
+        self.converter = converter
+        self.merger = merger
+        self.device = device
+        self.extract_hook = extract_hook
+        self.extract_func = extract_func
+        self.features = None
+
+    def get_iterator(self, name):
+        """Returns the iterator of the given name."""
+        return self._iterators[name]
+
+    def get_all_iterators(self):
+        """Returns a dictionary of all iterators."""
+        return dict(self._iterators)
+
+    def get_target(self, name):
+        """Returns the target link of the given name."""
+        return self._targets[name]
+
+    def get_all_targets(self):
+        """Returns a dictionary of all target links."""
+        return dict(self._targets)
+
+    def __call__(self, trainer=None):
+        """Executes the extractor extension.
+
+        Unlike usual extensions, this extension can be executed without passing
+        a trainer object.
+
+        Args:
+            trainer (~chainer.training.Trainer): Trainer object that invokes
+                this extension. It can be omitted in case of calling this
+                extension manually.
+
+        """
+        self.features = self.extract()
+        return self.features
+
+    def extract(self):
+        """Extractes features using the target model.
+
+        This method runs the extraction loop over the dataset. It accumulates
+        the extracted features to a :class:`~chainer.Variable`.
+
+        Users can override this method to customize the extraction routine.
+
+        Returns:
+            dict: Features variable.
+
+        """
+        iterator = self._iterators['main']
+        target = self._targets['main']
+        extract_func = self.extract_func or target
+
+        if self.extract_hook:
+            self.extract_hook(self)
+        it = copy.copy(iterator)
+
+        features = []
+        for batch in it:
+            in_arrays = self.converter(batch, self.device)
+            if isinstance(in_arrays, tuple):
+                in_vars = tuple(variable.Variable(x, volatile='on')
```

Comment:
What raise the `volatile` flag for the extracted features (also for the following two Variable creation locations)?